src/boot/libu64/gfxprint.c

```diff
@@ -0,0 +1,272 @@
+        gDPSetTileSize(this->gListp++, i * 2 + 1, 0, 0, (2 - 1) << G_TEXTURE_IMAGE_FRAC,
+                       (8 - 1) << G_TEXTURE_IMAGE_FRAC);
+    }
+}
+
+void gfxprint_color(gfxprint* this, u32 r, u32 g, u32 b, u32 a) {
+    this->color.r = r;
+    this->color.g = g;
+    this->color.b = b;
+    this->color.a = a;
+
+    gDPPipeSync(this->gListp++);
+    gDPSetColor(this->gListp++, G_SETPRIMCOLOR, this->color.rgba);
+}
+
+void gfxprint_locate(gfxprint* this, s32 x, s32 y) {
+    this->positionX = this->offsetX + x * 4;
+    this->positionY = this->offsetY + y * 4;
+}
+
+void gfxprint_locate8x8(gfxprint* this, s32 x, s32 y) {
+    gfxprint_locate(this, x * 8, y * 8);
+}
+
+void gfxprint_setoffset(gfxprint* this, s32 x, s32 y) {
+    this->offsetX = x * 4;
+    this->offsetY = y * 4;
+}
+
+void gfxprint_putc1(gfxprint* this, u8 c) {
+    u32 tile = (c & 0xFF) * 2;
+    u16 x0 = (c & 4);
+    u16 x1 = (c >> 3);
+
+    if (gfxprint_isChanged(this)) {
+        gfxprint_clrChanged(this);
+        gDPPipeSync(this->gListp++);
+
+        if (gfxprint_isGradient(this)) {
+            gDPSetTextureLUT(this->gListp++, G_TT_RGBA16);
+            gDPSetCycleType(this->gListp++, G_CYC_2CYCLE);
+            gDPSetRenderMode(this->gListp++, G_RM_PASS, G_RM_XLU_SURF2);
+            gDPSetCombineMode(this->gListp++, G_CC_INTERFERENCE, G_CC_PASS2);
+        } else {
+            gDPSetTextureLUT(this->gListp++, G_TT_IA16);
+            gDPSetCycleType(this->gListp++, G_CYC_1CYCLE);
+            gDPSetRenderMode(this->gListp++, G_RM_XLU_SURF, G_RM_XLU_SURF2);
+            gDPSetCombineMode(this->gListp++, G_CC_MODULATEIDECALA_PRIM, G_CC_MODULATEIDECALA_PRIM);
+        }
+    }
+
+    if (gfxprint_isShadow(this)) {
+
+        gDPSetColor(this->gListp++, G_SETPRIMCOLOR, 0);
+
+        if (gfxprint_isHighres(this)) {
+            gSPTextureRectangle(this->gListp++, (this->positionX + 4) << 1, (this->positionY + 4) << 1,
+                                (this->positionX + 4 + 32) << 1, (this->positionY + 4 + 32) << 1, tile, x0 << 6,
+                                x1 << 8, 512, 512);
+        } else {
+            gSPTextureRectangle(this->gListp++, this->positionX + 4, this->positionY + 4, this->positionX + 4 + 32,
+                                this->positionY + 4 + 32, tile, x0 << 6, x1 << 8, 1024, 1024);
+        }
+        gDPSetColor(this->gListp++, G_SETPRIMCOLOR, this->color.rgba);
+    }
+
+    if (gfxprint_isHighres(this)) {
+        gSPTextureRectangle(this->gListp++, this->positionX << 1, this->positionY << 1, (this->positionX + 32) << 1,
+                            (this->positionY + 32) << 1, tile, x0 << 6, x1 << 8, 512, 512);
+    } else {
+        gSPTextureRectangle(this->gListp++, this->positionX, this->positionY, this->positionX + 32,
+                            this->positionY + 32, tile, x0 << 6, x1 << 8, 1024, 1024);
+    }
+
+    this->positionX += 32;
+}
+
+void gfxprint_putc(gfxprint* this, u8 c) {
+    u8 param = c;
```

Comment:
Since those are actual C strings I think we should use `char` instead
```suggestion
void gfxprint_putc(gfxprint* this, char c) {
    char param = c;
```